File: python-sidecar/app/engines/video_generator/hyperframes/template_sets/v2_news_set.py
import os
import re

def render_scene_html(scene_id: str, sid: str, data: dict, theme: dict) -> str:
    """Render the inner HTML for V2 News Visual System scenes."""
    
    def escape(s):
        if not s: return ""
        return str(s).replace("&", "&amp;").replace("<", "&lt;").replace(">", "&gt;").replace('"', "&quot;")
    
    html = ""
    
    if sid == "hook":
        headline = escape(data.get("headline", ""))
        subhead = escape(data.get("subhead", ""))
        
        bg_html = f'<div class="bg gradient-news-dark"></div>'
        if data.get("bgSrc"):
            bg_html = f'<div class="bg kb-zoom-in" style="background-image: url(\'{data["bgSrc"]}\')"></div>'
            
        overlay_html = '<div class="overlay" style="opacity: 0.55"></div>'
        
        html = f'''{bg_html}
        {overlay_html}
        <div id="{scene_id}-hook-layout" class="layout-hook" style="position: absolute; inset: 0; z-index: 5; display: flex; flex-direction: column; align-items: center; justify-content: center; padding: 200px 80px 300px; text-align: center;">
            <div id="{scene_id}-headline" class="hook-headline shimmer-sweep-target">{headline}</div>
            <div id="{scene_id}-subhead" class="hook-subhead">{subhead}</div>
        </div>'''
        
    elif sid == "comparison":
        l_color = data.get("left", {}).get("color", "cyan")
        r_color = data.get("right", {}).get("color", "purple")
        r_winner = data.get("right", {}).get("winner", False)
        
        l_label = escape(data.get("left", {}).get("label", ""))
        l_value = escape(data.get("left", {}).get("value", ""))
        r_label = escape(data.get("right", {}).get("label", ""))
        r_value = escape(data.get("right", {}).get("value", ""))
        
        w_class = " card-winner" if r_winner else ""
        w_badge = '<div class="cmp-winner-badge">WINNER</div>' if r_winner else ""
        
        html = f'''<div id="{scene_id}-comp-layout" class="layout-comparison">
          <div id="{scene_id}-left" class="cmp-card cmp-left color-{l_color}">
            <div class="cmp-label">{l_label}</div>
            <div class="cmp-value">{l_value}</div>
          </div>
          <div id="{scene_id}-vs" class="cmp-vs">VS</div>
          <div id="{scene_id}-right" class="cmp-card cmp-right color-{r_color}{w_class}">
            <div class="cmp-label">{r_label}</div>
            <div class="cmp-value">{r_value}</div>
            {w_badge}
          </div>
        </div>'''
        
    elif sid == "stat-hero":
        val = escape(data.get("value", ""))
        lbl = escape(data.get("label", ""))
        ctx = escape(data.get("context", ""))
        
        ctx_html = f'<div id="{scene_id}-ctx" class="stat-context">{ctx}</div>' if ctx else ""
        
        html = f'''<div id="{scene_id}-stat-layout" class="layout-stat-hero">
          <div id="{scene_id}-val" class="stat-value shimmer-sweep-target">{val}</div>
          <div id="{scene_id}-lbl" class="stat-label">{lbl}</div>
          {ctx_html}
        </div>'''
        
    elif sid == "feature-list":
        title = escape(data.get("title", ""))
        bullets = ""
        for i, b in enumerate(data.get("bullets", [])):
            bullets += f'''
            <div id="{scene_id}-b{i}" class="feat-bullet feat-bullet-{i}">
              <div class="feat-dot"></div>
              <div class="feat-text">{escape(b)}</div>
            </div>'''
            
        html = f'''<div id="{scene_id}-feat-layout" class="layout-feature-list">
          <div id="{scene_id}-card" class="feat-card">
            <div class="feat-title">{title}</div>
            <div id="{scene_id}-rule" class="feat-rule"></div>
            <div class="feat-bullets">
              {bullets}
            </div>
          </div>
        </div>'''
        
    elif sid == "callout":
        tag = escape(data.get("tag", ""))
        statement = escape(data.get("statement", ""))
        
        tag_html = f'<div class="callout-tag">{tag}</div>' if tag else ""
        
        html = f'''<div id="{scene_id}-callout-layout" class="layout-callout">
          <div id="{scene_id}-card" class="callout-card">
            {tag_html}
            <div class="callout-statement">{statement}</div>
          </div>
        </div>'''
        
    elif sid == "outro":
        cta = escape(data.get("ctaTop", ""))
        ch = escape(data.get("channelName", ""))
        src = escape(data.get("source", ""))
        
        tt_card = f'''<div id="{scene_id}-tt-card" class="tt-card">
          <img class="tt-avatar" src="assets/avatar.png" alt="KNReup" crossorigin="anonymous" />
          <div class="tt-profile-info">
            <div class="tt-display-name">KNReup News</div>
            <div class="tt-handle">@knreup</div>
            <div class="tt-followers">1.2M followers</div>
          </div>
          <div id="{scene_id}-tt-follow-btn" class="tt-follow-btn">
            <span id="{scene_id}-tt-btn-follow" class="tt-btn-text">Follow</span>
            <span id="{scene_id}-tt-btn-following" class="tt-btn-text tt-btn-text-following">
              <span>Following</span>
              <span class="tt-check-icon"><svg viewBox="0 0 24 24" fill="none" stroke="#fff" stroke-width="2.5" stroke-linecap="round" stroke-linejoin="round"><polyline points="20 6 9 17 4 12"></polyline></svg></span>
            </span>
          </div>
        </div>'''
        
        html = f'''<div id="{scene_id}-outro-layout" class="layout-outro">
          <div id="{scene_id}-cta" class="out-cta-top">{cta}</div>
          <div id="{scene_id}-ch" class="out-channel">{ch}</div>
          <div id="{scene_id}-ul" class="out-underline"></div>
          <div id="{scene_id}-src" class="out-source">Nguồn: {src}</div>
        </div>
        {tt_card}'''
    elif sid == "quote":
        text = escape(data.get("text", ""))
        author = escape(data.get("author", ""))
        
        html = f'''<div id="{scene_id}-quote-layout" class="layout-quote">
          <div id="{scene_id}-mark" class="quote-mark">"</div>
          <div id="{scene_id}-txt" class="quote-text">{text}</div>
          <div id="{scene_id}-auth" class="quote-author">{author}</div>
        </div>'''
        
    elif sid == "timeline":
        title = escape(data.get("title", ""))
        rows = ""
        for i, ev in enumerate(data.get("events", [])):
            yr = escape(ev.get("year", ""))
            evt = escape(ev.get("event", ""))
            rows += f'''
            <div id="{scene_id}-r{i}" class="tl-row">
              <div class="tl-dot"></div>
              <div class="tl-year">{yr}</div>
              <div class="tl-event">{evt}</div>
            </div>'''
            if i < len(data.get("events", [])) - 1:
                rows += f'<div class="tl-line"></div>'
        
        html = f'''<div id="{scene_id}-tl-layout" class="layout-timeline">
          <div id="{scene_id}-t" class="tl-title">{title}</div>
          <div class="tl-track">{rows}
          </div>
        </div>'''
        
    elif sid == "countdown":
        label = escape(data.get("label", ""))
        number = escape(data.get("number", ""))
        unit = escape(data.get("unit", ""))
        
        html = f'''<div id="{scene_id}-cd-layout" class="layout-countdown">
          <div id="{scene_id}-lbl" class="cd-label">{label}</div>
          <div id="{scene_id}-num" class="cd-number">{number}</div>
          <div id="{scene_id}-unit" class="cd-unit">{unit}</div>
        </div>'''

    elif sid == "data-grid":
        title = escape(data.get("title", ""))
        cells = ""
        for i, c in enumerate(data.get("cells", [])):
            val = escape(c.get("value", ""))
            lbl = escape(c.get("label", ""))
            cells += f'''
              <div id="{scene_id}-c{i}" class="dg-cell">
                <div class="dg-cell-value">{val}</div>
                <div class="dg-cell-label">{lbl}</div>
              </div>'''
        
        html = f'''<div id="{scene_id}-dg-layout" class="layout-data-grid">
          <div id="{scene_id}-t" class="dg-title">{title}</div>
          <div class="dg-grid">{cells}
          </div>
        </div>'''

    elif sid == "title-card":
        category = escape(data.get("category", ""))
        title = escape(data.get("title", ""))
        date = escape(data.get("date", ""))
        
        html = f'''<div id="{scene_id}-tc-layout" class="layout-title-card">
          <div id="{scene_id}-cat" class="tc-category">{category}</div>
          <div id="{scene_id}-title" class="tc-title">{title}</div>
          <div id="{scene_id}-div" class="tc-divider"></div>
          <div id="{scene_id}-date" class="tc-date">{date}</div>
        </div>'''

    elif sid == "split-image":
        headline = escape(data.get("headline", ""))
        body = escape(data.get("body", ""))
        
        html = f'''<div id="{scene_id}-si-layout" class="layout-split-image">
          <div id="{scene_id}-img" class="split-img-zone"></div>
          <div class="split-text-zone">
            <div id="{scene_id}-hl" class="split-headline">{headline}</div>
            <div id="{scene_id}-bd" class="split-body">{body}</div>
          </div>
        </div>'''

    else:
        # Fallback — render as callout style
        text = escape(str(data.get("statement", data.get("text", data.get("headline", sid)))))
        html = f'''<div class="layout-callout">
          <div class="callout-card" style="opacity:1;">
            <div class="callout-statement">{text}</div>
          </div>
        </div>'''
    
    return html

def get_shell_html() -> str:
    """Return the persistent shell elements for the V2 News layout."""
    return '''
    <!-- Shell: persistent brand elements (no data-start -> always visible) -->
    <div class="shell-bg"></div>

    <div class="brand-shell-header">
      <div class="brand-icon">&gt;_</div>
      <div class="brand-text">
        <div class="brand-name">KNReup News</div>
        <div class="brand-tag">TIN CÔNG NGHỆ</div>
      </div>
    </div>

    <div class="brand-shell-handle">
      <span class="handle-music">&#9835;</span>
      <span class="handle-text">@knreup</span>
    </div>

    <div class="brand-shell-keyword">
      <span>vnexpress.net</span>
    </div>
    
    <!-- Grain overlay -->
    <div class="grain-overlay" style="background-image: url('data:image/svg+xml,%3Csvg viewBox=%220 0 200 200%22 xmlns=%22http://www.w3.org/2000/svg%22%3E%3Cfilter id=%22noiseFilter%22%3E%3CfeTurbulence type=%22fractalNoise%22 baseFrequency=%220.65%22 numOctaves=%223%22 stitchTiles=%22stitch%22/%3E%3C/filter%3E%3Crect width=%22100%25%22 height=%22100%25%22 filter=%22url(%23noiseFilter)%22 opacity=%220.35%22/%3E%3C/svg%3E');"></div>
    '''

def render_scene_animation(scene_id: str, sid: str, data: dict, start: float) -> str:
    """Return GSAP animation string for the given scene."""
    s = start
    js = ""
    
    if sid == "hook":
        js += f'tl.fromTo("#{scene_id}-headline", {{ scale: 0.5, opacity: 0 }}, {{ scale: 1, opacity: 1, duration: 0.6 }}, {s + 0.15});\n'
        js += f'tl.fromTo("#{scene_id}-subhead", {{ y: 60, opacity: 0 }}, {{ y: 0, opacity: 1, duration: 0.5 }}, {s + 0.55});\n'
    elif sid == "comparison":
        js += f'tl.fromTo("#{scene_id}-left", {{ x: -80, opacity: 0 }}, {{ x: 0, opacity: 1, duration: 0.5 }}, {s + 0.15});\n'
        js += f'tl.fromTo("#{scene_id}-vs", {{ scale: 0.5, opacity: 0 }}, {{ scale: 1, opacity: 1, duration: 0.35 }}, {s + 0.45});\n'
        js += f'tl.fromTo("#{scene_id}-right", {{ x: 80, opacity: 0 }}, {{ x: 0, opacity: 1, duration: 0.5 }}, {s + 0.6});\n'
    elif sid == "stat-hero":
        js += f'tl.fromTo("#{scene_id}-val", {{ scale: 0.4, opacity: 0 }}, {{ scale: 1, opacity: 1, duration: 0.6 }}, {s + 0.15});\n'
        js += f'tl.fromTo("#{scene_id}-lbl", {{ y: 40, opacity: 0 }}, {{ y: 0, opacity: 1, duration: 0.45 }}, {s + 0.55});\n'
        if data.get("context"):
            js += f'tl.fromTo("#{scene_id}-ctx", {{ y: 30, opacity: 0 }}, {{ y: 0, opacity: 1, duration: 0.4 }}, {s + 0.85});\n'
    elif sid == "feature-list":
        js += f'tl.fromTo("#{scene_id}-card", {{ y: 60, scale: 0.95, opacity: 0 }}, {{ y: 0, scale: 1, opacity: 1, duration: 0.5 }}, {s + 0.1});\n'
        js += f'tl.fromTo("#{scene_id}-rule", {{ scaleX: 0, opacity: 1 }}, {{ scaleX: 1, opacity: 1, duration: 0.4 }}, {s + 0.45});\n'
        for i in range(len(data.get("bullets", []))):
            js += f'tl.fromTo("#{scene_id}-b{i}", {{ x: -40, opacity: 0 }}, {{ x: 0, opacity: 1, duration: 0.4 }}, {s + 0.6 + i * 0.15});\n'
    elif sid == "callout":
        js += f'tl.fromTo("#{scene_id}-card", {{ y: 50, scale: 0.92, opacity: 0 }}, {{ y: 0, scale: 1, opacity: 1, duration: 0.55 }}, {s + 0.2});\n'
    elif sid == "outro":
        js += f'tl.fromTo("#{scene_id}-cta", {{ opacity: 0, y: -30 }}, {{ opacity: 1, y: 0, duration: 0.45 }}, {s + 0.2});\n'
        js += f'tl.fromTo("#{scene_id}-ch", {{ scale: 0.6, opacity: 0 }}, {{ scale: 1, opacity: 1, duration: 0.55 }}, {s + 0.55});\n'
        js += f'tl.fromTo("#{scene_id}-ul", {{ width: 0 }}, {{ width: "600px", duration: 0.5 }}, {s + 0.9});\n'
        js += f'tl.fromTo("#{scene_id}-src", {{ opacity: 0, y: 20 }}, {{ opacity: 1, y: 0, duration: 0.4 }}, {s + 1.3});\n'
        
        # TikTok card sequence
        ttBase = s + 1.6
        js += f'tl.fromTo("#{scene_id}-tt-card", {{ opacity: 0, y: 300, xPercent: -50 }}, {{ opacity: 1, y: 0, xPercent: -50, duration: 0.5 }}, {ttBase});\n'
        js += f'tl.to("#{scene_id}-tt-follow-btn", {{ scale: 0.92, duration: 0.15 }}, {ttBase + 0.9});\n'
        js += f'tl.to("#{scene_id}-tt-follow-btn", {{ scale: 1, duration: 0.4 }}, {ttBase + 1.05});\n'
        js += f'tl.to("#{scene_id}-tt-btn-follow", {{ opacity: 0, duration: 0.08 }}, {ttBase + 1.05});\n'
        js += f'tl.to("#{scene_id}-tt-btn-following", {{ opacity: 1, duration: 0.08 }}, {ttBase + 1.08});\n'
    elif sid == "quote":
        js += f'tl.fromTo("#{scene_id}-mark", {{ scale: 0.3, opacity: 0 }}, {{ scale: 1, opacity: 0.2, duration: 0.5 }}, {s + 0.1});\n'
        js += f'tl.fromTo("#{scene_id}-txt", {{ y: 40, opacity: 0 }}, {{ y: 0, opacity: 1, duration: 0.7 }}, {s + 0.3});\n'
        js += f'tl.fromTo("#{scene_id}-auth", {{ y: 20, opacity: 0 }}, {{ y: 0, opacity: 1, duration: 0.5 }}, {s + 0.8});\n'
    elif sid == "timeline":
        js += f'tl.fromTo("#{scene_id}-t", {{ y: -30, opacity: 0 }}, {{ y: 0, opacity: 1, duration: 0.5 }}, {s + 0.1});\n'
        for i in range(len(data.get("events", []))):
            js += f'tl.fromTo("#{scene_id}-r{i}", {{ x: -40, opacity: 0 }}, {{ x: 0, opacity: 1, duration: 0.4 }}, {s + 0.4 + i * 0.2});\n'
    elif sid == "countdown":
        js += f'tl.fromTo("#{scene_id}-lbl", {{ opacity: 0 }}, {{ opacity: 1, duration: 0.4 }}, {s + 0.1});\n'
        js += f'tl.fromTo("#{scene_id}-num", {{ scale: 2.5, opacity: 0 }}, {{ scale: 1, opacity: 1, duration: 1.0, ease: "elastic.out(1,0.4)" }}, {s + 0.3});\n'
        js += f'tl.fromTo("#{scene_id}-unit", {{ opacity: 0 }}, {{ opacity: 1, duration: 0.4 }}, {s + 0.8});\n'
    elif sid == "data-grid":
        js += f'tl.fromTo("#{scene_id}-t", {{ opacity: 0 }}, {{ opacity: 1, duration: 0.4 }}, {s + 0.1});\n'
        for i in range(len(data.get("cells", []))):
            js += f'tl.fromTo("#{scene_id}-c{i}", {{ scale: 0.85, opacity: 0 }}, {{ scale: 1, opacity: 1, duration: 0.5, ease: "back.out(1.5)" }}, {s + 0.3 + i * 0.15});\n'
    elif sid == "title-card":
        js += f'tl.fromTo("#{scene_id}-cat", {{ opacity: 0 }}, {{ opacity: 1, duration: 0.4 }}, {s + 0.1});\n'
        js += f'tl.fromTo("#{scene_id}-title", {{ y: 60, opacity: 0 }}, {{ y: 0, opacity: 1, duration: 0.8, ease: "power4.out" }}, {s + 0.3});\n'
        js += f'tl.fromTo("#{scene_id}-div", {{ scaleX: 0, opacity: 0 }}, {{ scaleX: 1, opacity: 1, duration: 0.5 }}, {s + 0.7});\n'
        js += f'tl.fromTo("#{scene_id}-date", {{ opacity: 0 }}, {{ opacity: 1, duration: 0.4 }}, {s + 1.0});\n'
    elif sid == "split-image":
        js += f'tl.fromTo("#{scene_id}-img", {{ y: -50, opacity: 0 }}, {{ y: 0, opacity: 1, duration: 0.7 }}, {s + 0.1});\n'
        js += f'tl.fromTo("#{scene_id}-hl", {{ y: 30, opacity: 0 }}, {{ y: 0, opacity: 1, duration: 0.6 }}, {s + 0.5});\n'
        js += f'tl.fromTo("#{scene_id}-bd", {{ opacity: 0 }}, {{ opacity: 1, duration: 0.5 }}, {s + 0.8});\n'
        
    return js

def get_css() -> str:
    """Return the CSS string for V2 News Visual System."""
    import os
    css_path = os.path.join(os.path.dirname(__file__), "v2_visual_styles.css")
    if os.path.exists(css_path):
        with open(css_path, "r", encoding="utf-8") as f:
            return f.read()
    return ""

File: src/components/TemplatePreview/sets/v2NewsSet.ts
import { SmartTemplate, ThemePalette, el } from '../templateData';
import type { TemplateSet } from './index';

// Helper to get persistent shell HTML
const getShellHtml = (d: any) => `
    <div class="shell-bg"></div>
    <div class="brand-shell-header">
      <div class="brand-icon">&gt;_</div>
      <div class="brand-text">
        <div class="brand-name">${d.channelName || 'KNReup News'}</div>
        <div class="brand-tag">TIN CÔNG NGHỆ</div>
      </div>
    </div>
    <div class="brand-shell-handle">
      <span class="handle-music">&#9835;</span>
      <span class="handle-text">@knreup</span>
    </div>
    <div class="brand-shell-keyword">
      <span>${d.source || 'vnexpress.net'}</span>
    </div>
`;

// ── V2 VISUAL SYSTEM TEMPLATES ──────────────────────────────────────────────
export const V2_NEWS_TEMPLATES: SmartTemplate[] = [
    {
        id: 'hook', name: 'Hook', icon: '⚡', desc: 'Headline + Ken Burns BG',
        sampleData: { 
            headline: 'TIẾN SĨ TOÁN GỐC NGA BIẾN AI THÀNH CỖ MÁY', 
            subhead: 'Công nghệ AI đang mở ra kỷ nguyên mới về...',
            bgSrc: 'https://images.unsplash.com/photo-1550751827-4bd374c3f58b?w=1080&h=1920&fit=crop'
        },
        render(c, d, theme) {
            c.style.padding = '0'; // reset default padding
            
            const bgHtml = d.bgSrc 
                ? `<div class="bg-blur" style="position: absolute; inset: 0; background-image: url('${d.bgSrc}'); background-size: cover; background-position: center; filter: blur(35px); opacity: 0.4; transform: scale(1.1); z-index: 0;"></div>
                   <div class="bg" style="background-image: url('${d.bgSrc}'); position: absolute; top: 50%; left: 50%; transform: translate(-50%, -50%); width: 1040px; height: 585px; background-size: cover; background-position: center; border-radius: 20px; box-shadow: 0 20px 40px rgba(0,0,0,0.4); z-index: 1;"></div>`
                : `<div class="bg gradient-news-dark" style="position: absolute; inset: 0; z-index: 1;"></div>`;
                
            c.innerHTML = getShellHtml(d) + `
                ${bgHtml}
                <div class="overlay" style="opacity: 0.1; position: absolute; inset: 0; background: #000; z-index: 2;"></div>
                <div class="layout-hook" style="position: absolute; inset: 0; z-index: 5; display: flex; flex-direction: column; align-items: center; justify-content: center; padding: 200px 80px 300px; text-align: center;">
                    <div class="hook-headline">${d.headline}</div>
                    ${d.subhead ? `<div class="hook-subhead">${d.subhead}</div>` : ''}
                </div>
            `;
        },
        animate(c, tl) {
            const h = c.querySelector('.hook-headline');
            const s = c.querySelector('.hook-subhead');
            if (h) tl.fromTo(h, { scale: 0.5, opacity: 0 }, { scale: 1, opacity: 1, duration: 0.6 }, 0.15);
            if (s) tl.fromTo(s, { y: 60, opacity: 0 }, { y: 0, opacity: 1, duration: 0.5 }, 0.55);
        }
    },
    {
        id: 'stat-hero', name: 'Stat Hero', icon: '📊', desc: 'Big number highlight',
        sampleData: { value: '97%', label: 'Người dùng hài lòng', context: 'Khảo sát Q4 2025' },
        render(c, d, theme) {
            c.style.padding = '0';
            c.innerHTML = getShellHtml(d) + `
                <div class="layout-stat-hero" style="position: absolute; inset: 0; z-index: 5; display: flex; flex-direction: column; align-items: center; justify-content: center; padding: 120px 80px;">
                    <div class="stat-value">${d.value}</div>
                    <div class="stat-label">${d.label}</div>
                    ${d.context ? `<div class="stat-context">${d.context}</div>` : ''}
                </div>
            `;
        },
        animate(c, tl) {
            const v = c.querySelector('.stat-value');
            const l = c.querySelector('.stat-label');
            const ctx = c.querySelector('.stat-context');
            if (v) tl.fromTo(v, { scale: 0.4, opacity: 0 }, { scale: 1, opacity: 1, duration: 0.6 }, 0.15);
            if (l) tl.fromTo(l, { y: 40, opacity: 0 }, { y: 0, opacity: 1, duration: 0.45 }, 0.55);
            if (ctx) tl.fromTo(ctx, { y: 30, opacity: 0 }, { y: 0, opacity: 1, duration: 0.4 }, 0.85);
        }
    },
    {
        id: 'comparison', name: 'Comparison', icon: '⚖️', desc: 'VS side-by-side',
        sampleData: { left: { label: 'iPhone 16', value: '$799', color: 'cyan' }, right: { label: 'Galaxy S25', value: '$749', color: 'purple', winner: true } },
        render(c, d, theme) {
            c.style.padding = '0';
            const lc = d.left.color || 'cyan';
            const rc = d.right.color || 'purple';
            const win = d.right.winner ? ' card-winner' : '';
            c.innerHTML = getShellHtml(d) + `
                <div class="layout-comparison" style="position: absolute; inset: 0; z-index: 5; display: flex; flex-direction: column; align-items: center; justify-content: center; padding: 100px 80px; gap: 40px;">
                    <div class="cmp-card cmp-left color-${lc}">
                        <div class="cmp-label">${d.left.label}</div>
                        <div class="cmp-value">${d.left.value}</div>
                    </div>
                    <div class="cmp-vs">VS</div>
                    <div class="cmp-card cmp-right color-${rc}${win}">
                        <div class="cmp-label">${d.right.label}</div>
                        <div class="cmp-value">${d.right.value}</div>
                        ${d.right.winner ? '<div class="cmp-winner-badge">WINNER</div>' : ''}
                    </div>
                </div>
            `;
        },
        animate(c, tl) {
            const leftCard = c.querySelector('.cmp-left');
            const vs = c.querySelector('.cmp-vs');
            const rightCard = c.querySelector('.cmp-right');
            if (leftCard) tl.fromTo(leftCard, { x: -80, opacity: 0 }, { x: 0, opacity: 1, duration: 0.5 }, 0.15);
            if (vs) tl.fromTo(vs, { scale: 0.5, opacity: 0 }, { scale: 1, opacity: 1, duration: 0.35 }, 0.45);
            if (rightCard) tl.fromTo(rightCard, { x: 80, opacity: 0 }, { x: 0, opacity: 1, duration: 0.5 }, 0.6);
        }
    },
    {
        id: 'feature-list', name: 'Feature List', icon: '📋', desc: 'Title + bullet points',
        sampleData: { title: 'Tính năng nổi bật', bullets: ['AI tự động phân tích', 'Render real-time 4K', 'Export đa nền tảng', 'Chỉnh sửa không phá hủy'] },
        render(c, d, theme) {
            c.style.padding = '0';
            const bulletsHtml = d.bullets.map((b: string, i: number) => `
                <div class="feat-bullet feat-bullet-${i}">
                    <div class="feat-dot"></div>
                    <div class="feat-text">${b}</div>
                </div>
            `).join('');

            c.innerHTML = getShellHtml(d) + `
                <div class="layout-feature-list" style="position: absolute; inset: 0; z-index: 5; display: flex; align-items: center; justify-content: center; padding: 120px 80px;">
                    <div class="feat-card">
                        <div class="feat-title">${d.title}</div>
                        <div class="feat-rule"></div>
                        <div class="feat-bullets">
                            ${bulletsHtml}
                        </div>
                    </div>
                </div>
            `;
        },
        animate(c, tl) {
            const card = c.querySelector('.feat-card');
            const rule = c.querySelector('.feat-rule');
            const bullets = c.querySelectorAll('.feat-bullet');
            
            if (card) tl.fromTo(card, { y: 60, scale: 0.95, opacity: 0 }, { y: 0, scale: 1, opacity: 1, duration: 0.5 }, 0.1);
            if (rule) tl.fromTo(rule, { scaleX: 0, opacity: 1 }, { scaleX: 1, opacity: 1, duration: 0.4 }, 0.45);
            bullets.forEach((b, i) => {
                tl.fromTo(b, { x: -40, opacity: 0 }, { x: 0, opacity: 1, duration: 0.4 }, 0.6 + i * 0.15);
            });
        }
    },
    {
        id: 'callout', name: 'Callout', icon: '🔔', desc: 'Bold statement + tag',
        sampleData: { tag: 'CẢNH BÁO', statement: 'Hệ thống thanh toán sẽ ngừng hoạt động vào ngày 15/12' },
        render(c, d, theme) {
            c.style.padding = '0';
            c.innerHTML = getShellHtml(d) + `
                <div class="layout-callout" style="position: absolute; inset: 0; z-index: 5; display: flex; align-items: center; justify-content: center; padding: 100px 80px;">
                    <div class="callout-card">
                        ${d.tag ? `<div class="callout-tag">${d.tag}</div>` : ''}
                        <div class="callout-statement">${d.statement}</div>
                    </div>
                </div>
            `;
        },
        animate(c, tl) {
            const card = c.querySelector('.callout-card');
            if (card) tl.fromTo(card, { y: 50, scale: 0.92, opacity: 0 }, { y: 0, scale: 1, opacity: 1, duration: 0.55 }, 0.2);
        }
    },
    {
        id: 'outro', name: 'Outro', icon: '🎬', desc: 'CTA + follow card',
        sampleData: { ctaTop: 'THEO DÕI NGAY', channelName: 'KNReup News', source: 'vnexpress.net' },
        render(c, d, theme) {
            c.style.padding = '0';
            c.innerHTML = getShellHtml(d) + `
                <div class="layout-outro" style="position: absolute; inset: 0; z-index: 5; display: flex; flex-direction: column; align-items: center; padding: 300px 80px 200px;">
                    <div class="out-cta-top">${d.ctaTop}</div>
                    <div class="out-channel">${d.channelName}</div>
                    <div class="out-underline"></div>
                    <div class="out-source">Nguồn: ${d.source}</div>
                </div>
                <div id="tt-card" class="tt-card">
                    <img class="tt-avatar" src="assets/avatar.png" alt="Avatar" crossorigin="anonymous" />
                    <div class="tt-profile-info">
                        <div class="tt-display-name">${d.channelName}</div>
                        <div class="tt-handle">@knreup</div>
                        <div class="tt-followers">1.2M followers</div>
                    </div>
                    <div id="tt-follow-btn" class="tt-follow-btn">
                        <span id="tt-btn-follow" class="tt-btn-text">Follow</span>
                        <span id="tt-btn-following" class="tt-btn-text tt-btn-text-following" style="opacity: 0; position: absolute;">
                            <span>Following</span>
                            <span class="tt-check-icon"><svg viewBox="0 0 24 24" fill="none" stroke="#fff" stroke-width="2.5" stroke-linecap="round" stroke-linejoin="round" style="width: 24px; height: 24px;"><polyline points="20 6 9 17 4 12"></polyline></svg></span>
                        </span>
                    </div>
                </div>
            `;
        },
        animate(c, tl) {
            const cta = c.querySelector('.out-cta-top');
            const channel = c.querySelector('.out-channel');
            const underline = c.querySelector('.out-underline');
            const source = c.querySelector('.out-source');
            
            if (cta) tl.fromTo(cta, { opacity: 0, y: -30 }, { opacity: 1, y: 0, duration: 0.45 }, 0.2);
            if (channel) tl.fromTo(channel, { scale: 0.6, opacity: 0 }, { scale: 1, opacity: 1, duration: 0.55 }, 0.55);
            if (underline) tl.fromTo(underline, { width: 0 }, { width: "600px", duration: 0.5 }, 0.9);
            if (source) tl.fromTo(source, { opacity: 0, y: 20 }, { opacity: 1, y: 0, duration: 0.4 }, 1.3);

            const ttCard = c.querySelector('#tt-card');
            const ttBtn = c.querySelector('#tt-follow-btn');
            const ttFollow = c.querySelector('#tt-btn-follow');
            const ttFollowing = c.querySelector('#tt-btn-following');
            const ttBase = 1.6;

            if (ttCard) tl.fromTo(ttCard, { opacity: 0, y: 300, xPercent: -50 }, { opacity: 1, y: 0, xPercent: -50, duration: 0.5 }, ttBase);
            if (ttBtn) {
                tl.to(ttBtn, { scale: 0.92, duration: 0.15 }, ttBase + 0.9);
                tl.to(ttBtn, { scale: 1, duration: 0.4 }, ttBase + 1.05);
            }
            if (ttFollow) tl.to(ttFollow, { opacity: 0, duration: 0.08 }, ttBase + 1.05);
            if (ttFollowing) tl.to(ttFollowing, { opacity: 1, duration: 0.08 }, ttBase + 1.08);
            if (ttCard) tl.to(ttCard, { scale: 1.08, duration: 2.0 }, ttBase + 1.3); // Slow zoom
        }
    }
];

export const V2_NEWS_SET: TemplateSet = {
    id: 'v2-news',
    name: 'V2 Visual System',
    description: 'Advanced Tech News Style (Repo Test template)',
    templates: V2_NEWS_TEMPLATES,
};
